### PR TITLE
Add EqualityComparer<T>.Create

### LINKED ISSUE #24460: Add "Create" method to EqualityComparer<> class
EDITED by @stephentoub on 4/26/2022 to update proposal:
```diff
namespace System.Collections.Generic
{
    public abstract class EqualityComparer<T> : IEqualityComparer<T>, IEqualityComparer
    {
+        public static EqualityComparer<T> Create(Func<T?, T?, bool> equals, Func<T, int> getHashCode);
    }
}
```

To discuss:
- Making `getHashCode` optional such that the returned instance's GetHashCode would throw `NotSupportedException` if it's used.

---

_From @ViIvanov on December 14, 2017 20:15_

Now we have a `Create` method in [Comparer<> class](https://github.com/dotnet/coreclr/blob/master/src/mscorlib/src/System/Collections/Generic/Comparer.cs#L21):
```cs
public static Comparer<T> Create(Comparison<T> comparison)
```

I think, it will be usable to have an analog in EqualityComparer<>:
```cs
public static EqualityComparer<T> Create(Func<T, T, bool> equals, Func<T, int> getHashCode)
```

_Copied from original issue: dotnet/coreclr#15526_

### PR DESCRIPTION
**Pull request by stephentoub**
Fixes https://github.com/dotnet/runtime/issues/24460

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/Collections/Generic/EqualityComparer.cs [hunk 1]
```diff
@@ -62,6 +83,21 @@ internal virtual int LastIndexOf(T[] array, T value, int startIndex, int count)
 #endif
     }
 
+    internal sealed class DelegateEqualityComparer<T> : EqualityComparer<T>
```
Comment (ViIvanov) [inline src/libraries/System.Private.CoreLib/src/System/Collections/Generic/EqualityComparer.cs:86]: Why not to override `Equals(object)` and `GetHashCode()` to be able compare equality comparers?
Delegates are equality comparable, so it is possible to do.
Some parts of BCL uses comparing comparers to run more optimal code.

Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/Collections/Generic/EqualityComparer.cs:86]: Added.

**Review by eiriktsarpalis — APPROVED** [review]
Thanks!

**Comment by stephentoub** [comment]
failure is https://github.com/dotnet/runtime/issues/75391

**Pull request merged**